src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -236,32 +280,40 @@ protected void logEventAsync(final String eventType, JSONObject eventProperties,
         runOnLogThread(new Runnable() {
             @Override
             public void run() {
-                logEvent(eventType, copyEventProperties, apiProperties, timestamp, checkSession);
+                logEvent(eventType, copyEventProperties, apiProperties, timestamp, outOfSession);
             }
         });
     }
 
     protected long logEvent(String eventType, JSONObject eventProperties,
-            JSONObject apiProperties, long timestamp, boolean checkSession) {
+            JSONObject apiProperties, long timestamp, boolean outOfSession) {
         Log.d(TAG, "Logged event to Amplitude: " + eventType);
 
         if (optOut) {
             return -1;
         }
-        if (checkSession) {
-            startNewSessionIfNeeded(timestamp);
+
+        // skip session check if logging start_session or end_session events
+        boolean loggingSessionEvent = trackingSessionEvents &&
+                (eventType.equals(START_SESSION_EVENT) || eventType.equals(END_SESSION_EVENT));
+
+        if (!loggingSessionEvent && !outOfSession) {
+            // default case + corner case when async logEvent between onPause and onResume
+            if (!usingForegroundTracking || !inForeground){
```

Comment:
I think the !usingForegroundTracking is actually not necessary, since if we're not usingForegroundTracking, inForeground will always be false.